Properly reset default layer when resetting EEPROM

drashna opened the pull request:
## Description

When resetting the EEPROM, it should forcibly set the default layer. 

We were doing that by setting it to zero, but that doesn't trigger any of the normal checks.   This fixes that, and calls the proper function

## Types of Changes
- [x] Core
- [x] Bugfix
- [x] Enhancement/optimization

## Checklist
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented:
Test is erroring out, and no idea why. 

IIRC, this is why I didn't do this before....

Specifically, it looks like the "basic" test if failing

noroadsleft commented:
> Test is erroring out, and no idea why.
> 
> IIRC, this is why I didn't do this before....
> 
> Specifically, it looks like the "basic" test if failing

This looks to be something with `lib/googletest` submodule, which is returning [`0 FAILED TESTS` on this PR](https://travis-ci.org/qmk/qmk_firmware/builds/545390759#L1391-L1396) even though all 13 of the individual tests are passing.

Compare that output with that from #6124, which also passes all 13 tests, but [doesn't return any failure info](https://travis-ci.org/qmk/qmk_firmware/builds/544931682#L1384-L1386).

This may be related to our recent integration of Python, as the error, near as I can tell, is being [returned by a Python file](https://github.com/google/googletest/blob/ec44c6c1675c25b9827aacd08c02433cccde7780/googletest/test/gtest_output_test.py#L159-L160)

drashna commented:
Nope, it's not related to that.  

It's been a long standing issue with this.  I just don't know of a good solution.

fauxpark commented:
I suspect this has something to do with the addition of `keyboard_init()`:

https://github.com/qmk/qmk_firmware/blob/1cd26607bd4bd5bd718e5c1e2e3884fbcdf50f5d/tests/test_common/test_fixture.cpp#L23-L27

This function sets up the expectation of the mock `send_keyboard()` to be called exactly once, then calls `keyboard_init()`. That in turn calls `matrix_init()`, which in turn calls `matrix_init_quantum()`, which in turn calls `eeconfig_init()`, which in turn calls... `eeconfig_init_quantum()`.

The error claims that `send_keyboard_mock()` is called twice, but that kinda looks like an infinite loop to me. Evidently somewhere in there something is trying to send a keyboard report, and ordinarily this is fine as `keyboard_init()` would only be called once.

drashna commented:
The problem is that other methods don't work either. Such as including "action_layer.h", and calling `default_layer_set`.  

As for the loop, it doesn't happen.  `eeconfig_init()` is only called if the EEPROM is not in a valid state. And the first thing `eeconfig_init` does is set the state to be valit.  So it shouldn't get called again.   (but yes, it could lead to a loop, otherwise).

The problem is, no matter what I do, the only thing that works is a "soft set" of the default layer.  And that's not good enough.  

In fact, `keyboard_init` is closer to what we want to happen, since it will reset all the features, as well. As is, right now, everything just stays the same until you power cycle the board. Which is not a proper experience.

fauxpark commented:
The reason `default_layer_set()` also doesn't work is because it calls `default_layer_state_set()`, which eventually also leads to `send_keyboard_report()` through `clear_keyboard_but_mods_and_keys()`.

fauxpark commented:
I set up docker and ran `./util/travis_test.sh`, the above passes but I'm not sure if the keyboard clearing is necessary.

drashna commented:
Well, at the very least, fixed in the hackiest way possible.  Moved the functions to quantum.c, instead... as they probably should have been there from the start, and only call keyboard_init on avr/arm....

fauxpark commented:
So the keyboard clearing is needed then?
With this hack, the failing test is basically bypassed, and on a real board `send_keyboard()` will still be called twice :/

drashna commented:
Yes, I feel that it does, personally. 

Namely, settings like RGB don't get updated until you power cycle the board, or change the settings.  Same thing with the default layer.  Anything that read from a data structure instead of the EEPROM directly, which I think is EVERYTHING that uses EEPROM.

So, you'd reset the eeprom and then have to reset the the board.  

Ideally, we need to re-init everything, so that settings get freshly pulled from the eeprom.  
There are a couple of ways that we could do this.  This is the simplest. 
But we could call the init functions for everything from another helper function.... but this is re-inventing `keyboard_init`. 

The other way to do this, would be to soft reset the keyboard, but not jump to bootloader.   I don't know that we have a way to do this, at least implemented yet.   That would be the best option, IMO.  But that's beyond my skill level. 


As for calling `send_keyboard();` twice, I don't think that should be an issue.  IIRC, the report should be cleared when it's sent, so you shouldn't get "stuck", anywhere. 

And I've been running this on my keyboard for a bit now (calling `keyboard_init` from `eeconfig_user`), and haven't had issues.




> So the keyboard clearing is needed then?

For setting the default layer?   I think it is.  I know that for changing normal layers, it's a pain, but I don't think we have the detection and handling for that for default layers, so keys could get stuck.

fauxpark commented:
I'm not worried about the duplicate call breaking things, it just seems superfluous to me. But if it's unavoidable, why not simply change the test to expect two calls? Then we don't need the ifdef.

drashna commented:
because I have no idea how to do so.

fauxpark commented:
There's an example here: https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md#general-syntax

In test_fixture.cpp:
```cpp
EXPECT_CALL(driver, send_keyboard_mock(_)).Times(2);
```

😁

drashna commented:
Well, I could do that ....  My only concern is that it may break other things, or not detect them. 

But yeah, is seems to fix the issue.

fauxpark commented:
I'm not sure I follow...
Correct me if I'm wrong, but the ifdef is basically saying, "unless this is a real board (ie. not Travis/gmock), don't call this function here". The code is now aware that it's being tested, and will do something different if it is. If your fears that it might break something are correct, it will only be observed on a real board.

drashna commented:
Honestly, I think it's mostly paranoia about messing with code that I'm not familiar with.

drashna commented:
ALl this work, and I didn't think of just adding `keyboard_init()` to the `EEPROM_RESET` keycode ..... 

Since that's the only place it's actually really needed ........

fauxpark commented:
It's not needed for Bootmagic?

drashna commented:
Yes and no.  Lite doesn't need to.  

But bootmagic full, shoot, yeah it does.  But that's only because matrix init is called first....
But that is an easy fix: Call matrix_init again.

zvecr commented on tmk_core/common/bootmagic.c:
```diff
@@ -39,6 +39,7 @@ void bootmagic(void) {
     /* eeconfig clear */
     if (bootmagic_scan_keycode(BOOTMAGIC_KEY_EEPROM_CLEAR)) {
         eeconfig_init();
+        matrix_init();
```
Calling `matrix_init` here and `keyboard_init` when handling `EEPROM_RESET` feels inconsistent. Architecturally, invoking a core part of the init path outside the init path without any form of deinit sounds like the wrong thing to do.

For instance, what happens with assumptions that `matrix_init_kb` or `matrix_init_user` is only called on startup? (maybe configuring a piece of hardware that cannot be reconfigured later, allocating memory. Not 100% on an existing use case at the moment)

drashna replied:
I definitely agree with you on the inconsistency. 

However, calling `keyboard_init` here would lead to a situation where the keyboard could actually get stuck in an infinite loop, since bootmagic full is called from keyboard_init

Also, to muddle things, where things are init-ed is very inconsistent, which makes it hard to use just a single function for this, without changing a lot.   

Admittedly, this is putting off tech debt, but short of a massive change, (which I think needs to happen anyways), I'm not sure of a good way to implement that here.


but leaving the keyboard_init in the eeconfig_init isn't a good solution, either.  It has some ... weird behavior.  I've seen my boards (which were using this code) re-init the keyboard 2, 3, and sometimes 4 times, because of it. This is a lighter touch, as it shouldn't trigger it everywhere.

zvecr replied:
>Admittedly, this is putting off tech debt, but short of a massive change, 

Its worth noting that this is still adding to that technical debt, now the init order is muddy.

matrix_init being called twice, calls debounce_init again, which can do https://github.com/qmk/qmk_firmware/blob/7372ce6afdb1030463e09cef92e0d7befbd1d474/quantum/debounce/eager_pr.c#L44 without freeing the memory. Thats quite bad....

drashna commented:
Honestly, I think I'm going to just close this.  

I think the idea behind it is the right direction, but I think we need a lot of cleanup of code to ensure that there is no issue with doing so, rather than blindly re-init-ing things.
